Recommender.py for FastAPI

jpjon opened the pull request:
## Describe your changes
Created the `.py` file for the recommender that FastAPI will reference to.


## Issue number
#99 
Closes #99 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)


<!-- readthedocs-preview ploomber-sql start -->
----
:books: Documentation preview :books:: https://ploomber-sql--100.org.readthedocs.build/en/100/

<!-- readthedocs-preview ploomber-sql end -->

jpjon commented on mini-projects/movie-rec-system/movie_rec_system/recommendation/recommender.py:
```diff
@@ -0,0 +1,77 @@
+)
+
+
+def connect_duckdb():
```
Does this function depend on where/how we store the data?

For instance, if it's in a S3 bucket that we'll access through Lambda, wouldn't we have to adjust this function to account for that?

Because the bucket isn't yet set up, should I leave this as is for now?

lfunderburk replied:
Can you build this function `connect_duckdb` assuming you run the fastapi instance locally? 

Let's keep things simple

1. First get a local fastapi instance working
2. we upgrade or add new docker container for it
3. then we can modify the functionality to account for aws architecture
4. Add docstrings to your functions
5. Begin writing simple tests to ensure your functions work appropriately

jpjon replied:
This function has been renamed to `get_data` and works appropriately. 

https://github.com/ploomber/sql/pull/100/files#diff-9b1926a7b439d1798b4654dd83de94bab22f71c9bea3f3db147cde8e9f062b85R14-R26

jpjon commented:
Hi @lfunderburk 

Can you take a look at what I have so far? My main concern regards my comment on how we would connect to the S3 Bucket using Lambda.

Besides that, I think the main `get_recommendation` function is of viable use for FastAPI.

lfunderburk commented on mini-projects/movie-rec-system/movie_rec_system/recommendation/recommender.py:
```diff
@@ -0,0 +1,77 @@
+    pass
+
+
+def create_combined(df, weight=2):
```
please add more context
df could be anything

add a docstring and a sample of what df looks like

jpjon replied:
Docstring has been added: 
https://github.com/ploomber/sql/pull/100/files#diff-9b1926a7b439d1798b4654dd83de94bab22f71c9bea3f3db147cde8e9f062b85R30-R63

jpjon commented:
Please navigate to the `/recommendations` folder and run `uvicorn app:app --reload` to launch the API.

Go to `http://127.0.0.1:8000/docs#/default` in your browser to test out the API with a movie. I used the movie "Barbie" and received the expected output.

lfunderburk commented:
Hi @jpjon 

Did a bit of code clean up, type checking, pydantic modelling and added some testing

To execute tests

```
cd sql/mini-projects/movie-rec-system
pytest
```

From the Docker container we don't want to specify the path, we should be running things from the root directory (in this case `sql/mini-projects/movie-rec-system`)

Please execute the app as follows

```
 uvicorn movie_rec_system.app.app:app --reload
```

lfunderburk reviewed (changes requested):
To do:

1. Run the tests, there are one or two warnings. See if you can fix them
2. Add a test under `test/test_app.py` to check that a given movie is returning the desired result
3. Think of other ways the user might interact with the api - does it make sense to add another entry point?

jpjon commented:
Changed made to fix warnings:

Replaced @validator with @field_validator: https://events.bizzabo.com/458007/settings/registration/1eaNwY9AETInoqcCd_pBC3mdpyVozaTeGSNDrqw3yMsJPmiZu8MIS4Fbc0Qmsa1f

Added additional test for validating response:
https://events.bizzabo.com/458007/settings/registration/1eaNwY9AETInoqcCd_pBC3mdpyVozaTeGSNDrqw3yMsJPmiZu8MIS4Fbc0Qmsa1f

lfunderburk reviewed (approved):
Test coverage passes with new upgrades to pyproject.toml and app.py

Awesome!
